### ISSUE TITLE
What does GCAE converge to? A NMSE of 1?

### ISSUE THREAD

**Issue by richelbilderbeek** [richelbilderbeek/nsphs_ml_qt#61]
Currently, all runs converge to a normalized mean squared error of one. Ideally, this error should be zero.

These are both from run #50:

![nmse_in_time](https://user-images.githubusercontent.com/2098230/176421362-0b5f0058-ddb6-4ad9-81c7-a0de02aa4929.png)

![nmse_in_time](https://user-images.githubusercontent.com/2098230/176421271-fc5f3bda-3131-4292-b70f-4ad4fce907ed.png)

Note that sometimes NMSE goes down, see the top-left panel of run #42:

![issue_42_nmses_in_time](https://user-images.githubusercontent.com/2098230/176421717-55d5daa9-3754-4f87-b5c1-6f0b9451a9e2.png)

It feels worth exploring if maybe GCAE is trained to have a NMSE of 1, i.e. its highest fitness is when NMSE is 1, instead of 0.

**Comment by richelbilderbeek** [comment]
The `phenoloss` appears never to be used relevantly:

```
richel@N141CU:~/GitHubs/gcae$ egrep -R "phenoloss" --include=*.py
run_gcae.py:		phenoloss = loss_value
run_gcae.py:		phenoloss, phenoalpha = (0.,0.)
run_gcae.py:	tf.print(loss_value, other_loss4, other_loss3, other_loss2, other_loss, phenoloss, full_loss, alpha4, alpha3, alpha2, alpha, phenoalpha)
```

It seems the `gradientspheno` is more relevant:

```
	def combine(gradients, gradients2):
        [...]
	if phenomodel is not None:
		gradients3, phenoalpha = combine(gradients3, gradientspheno)
	else:
		phenoloss, phenoalpha = (0.,0.)
        [...]
	tf.print(loss_value, other_loss4, other_loss3, other_loss2, other_loss, phenoloss, full_loss, alpha4, alpha3, alpha2, alpha, phenoalpha)
	return loss_value
```

**Comment by richelbilderbeek** [comment]
As an error function, the squared errors are used.

However, most work is done by the gradients, e.g. `gradients3, phenoalpha = combine(gradients3, gradientspheno)`. I cannot verify that `combine` works correctly.

I will add the amazement of NMSE converging to 1 to the article.

**Comment by richelbilderbeek** [comment]
I checked by copying the values from `phenotype_predictions.csv` and doing the same in Calc. Also here, the error `=(C871-D871)^2`, then taking the average of the sum `=SUM(F2:F871)/870` resulted in the same value.

**Comment by richelbilderbeek** [comment]
I have checked that the phenotype file is correct ...

````
[richel@sens2021565-bianca data_issue_50_100]$ ls *.phe
data_issue_50_100.phe
```

I have checked the predicted values are correct

```
[richel@sens2021565-bianca data_issue_50_100]$ cat 10000.phe
```

So, input is correct.

**Comment by richelbilderbeek** [comment]
I have made a debug `.sif`, running it for setting 5:

```
[richel@sens2021565-bianca nsphs_ml_qt]$ mv nsphs_ml_qt_debug.sif nsphs_ml_qt.sif
[richel@sens2021565-bianca nsphs_ml_qt]$ cd ..
[richel@sens2021565-bianca ~]$ ls
98_clean_bianca.sh  bin  n_jobs.txt  nsphs_ml_qt  nsphs_ml_qt_results  README.md  richel-sens2021565  script.R
[richel@sens2021565-bianca ~]$ ./nsphs_ml_qt/scripts_bianca/
20_start_issue_28.sh         21_create_issue_5_params.sh  29_zip_issue_42.sh
20_start_issue_29.sh         22_create_issue_28_data.sh   29_zip.sh
20_start_issue_42_again.sh   22_create_issue_29_data.sh   91_poll_jobs.sh
20_start_issue_42.sh         22_create_issue_42_data.sh   92_poll_n_jobs.sh
20_start_issue_50.sh         22_create_issue_50_data.sh   98_clean_bianca.sh
20_start_issue_5.sh          22_create_issue_5_data.sh    
[richel@sens2021565-bianca ~]$ ./nsphs_ml_qt/scripts_bianca/20_start_issue_5.sh 
Starting time: 2022-07-04T09:43:36+0200
Running on computer with HOSTNAME: sens2021565-bianca.uppmax.uu.se
Running at location /home/richel
gcae_experiment_params_filename: /proj/sens2021565/nobackup/nsphs_ml_qt_results/data_issue_5_1/experiment_params.csv
unique_id: issue_5
jobid_21: 28771
jobid_22: 28772
jobid_24: 28773
jobid_25: 28774
jobid_26: 28775
jobid_29: 28776
End time: 2022-07-04T09:43:37+0200
Duration: 1 seconds
```

Goals is to detect a negative predicted value in the 25_ logs.

**Comment by richelbilderbeek** [comment]
Need to increase the verbosity then in `gcaer` as well...

**Comment by richelbilderbeek** [comment]
There goes:

```
[richel@sens2021565-bianca ~]$ ./nsphs_ml_qt/scripts_bianca/20_start_issue_5.sh 
Starting time: 2022-07-04T11:32:16+0200
Running on computer with HOSTNAME: sens2021565-bianca.uppmax.uu.se
Running at location /home/richel
gcae_experiment_params_filename: /proj/sens2021565/nobackup/nsphs_ml_qt_results/data_issue_5_1/experiment_params.csv
unique_id: issue_5
jobid_21: 28779
jobid_22: 28780
jobid_24: 28781
jobid_25: 28782
jobid_26: 28783
jobid_29: 28784
End time: 2022-07-04T11:32:17+0200
Duration: 1 seconds
```

**Comment by richelbilderbeek** [comment]
Folders are clean:

```
[richel@sens2021565-bianca ~]$ cd nsphs_ml_qt_results/
[richel@sens2021565-bianca nsphs_ml_qt_results]$ ls
[richel@sens2021565-bianca nsphs_ml_qt_results]$ ls
```

**Comment by richelbilderbeek** [comment]
Alrighty, I am going to use another approach, using simulated data.

**Comment by richelbilderbeek** [comment]
Running with bigger values now, as the result was yet inconclusive.

![Screenshot from 2022-07-04 17-19-26](https://user-images.githubusercontent.com/2098230/177182884-80cee920-116e-41cc-bb86-8f346fa08b6e.png)

**Comment by richelbilderbeek** [comment]
It seems more GenoCAE struggles with larger values ...?

![Screenshot from 2022-07-04 19-25-05](https://user-images.githubusercontent.com/2098230/177198176-09e0f915-557d-4d66-993d-e98beb6e5353.png)

**Comment by richelbilderbeek** [comment]
This test made the last picture:

```
test_that("use", {
  # If running with a simple additive trait, it should
  # not matter much if the phenotype value is -100, 100 or 200.
  #
  base_phenotype_values <- c(-100, 100, 200)
  results_list <- list()
  for (i in seq_along(base_phenotype_values)) {
    message(i)
    base_phenotype_value <- base_phenotype_values[i]
    assoc_qt_data <- plinkr::create_demo_assoc_qt_data(
      n_individuals = 100,
      traits = plinkr::create_additive_trait(
        mafs = 0.49,
        n_snps = 1,
        base_phenotype_value = base_phenotype_value,
        phenotype_increase = 0.5
      )
    )
    assoc_qt_data$data <- plinkr::convert_plink_text_data_to_plink_bin_data(
      assoc_qt_data$data
    )
    datadir <- paste0(
      file.path(gcaer::get_gcaer_tempfilename(), "issue_61"),
      "/"
    )
    data <- "issue_61_data"
    base_input_filename <- paste0(datadir, data)
    plinkr::save_plink_bin_data(
      plink_bin_data = assoc_qt_data$data,
      base_input_filename = base_input_filename
    )
    plinkr::save_phe_table(
      phe_table = assoc_qt_data$phenotype_data$phe_table,
      phe_filename = paste0(base_input_filename, ".phe")
    )

    gcae_experiment_params <- gcaer::create_gcae_experiment_params(
      gcae_setup = gcaer::create_test_gcae_setup(
        datadir = datadir,
        data = data,
        superpops = "",
        model_id = "M1",
        pheno_model_id = "p1"
      ),
      analyse_epochs = c(100, 200, 400, 800, 1600, 3200),
      metrics = ""
    )
    results <- gcaer::do_gcae_experiment(gcae_experiment_params)

    results$nmse_in_time_table$base_phenotype_value <- base_phenotype_value
    results_list[[i]] <- results$nmse_in_time_table
  }
  t <- dplyr::bind_rows(results_list)
  t$base_phenotype_value <- as.factor(t$base_phenotype_value)
  ggplot2::ggplot(
    t,
    ggplot2::aes(x = epoch, y = nmse, color = base_phenotype_value)
  ) + ggplot2::geom_line()

  ggplot2::ggsave("~/issue_61.png", width = 7, height = 7)
})
```

**Comment by richelbilderbeek** [comment]
I am going to run this on Rackham, to be able to see a clear difference. AFAICS, GCAE can handle negative numbers fine.

**Comment by richelbilderbeek** [comment]
Running:

```
[richel@rackham2 ~]$ sbatch nsphs_ml_qt/scripts_local/issue_61/start_issue_61.sh 
Submitted batch job 28166624
[richel@rackham2 ~]$ squeue -u $USER
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
          28166624      core issue_61   richel  R       0:00      1 r1224

```

**Comment by richelbilderbeek** [comment]
```
[richel@rackham2 ~]$ cat issue_61.log
base_input_filename: 
n_individuals: 
n_traits: 
n_snps_per_trait: 
singularity_filename: 
Starting time: 2022-07-05T10:27:01+0200
Running on computer with HOSTNAME: r1224
Running at location /home/richel
'nsphs_ml_qt.sif' running with arguments 'Rscript nsphs_ml_qt/scripts_local/issue_61/issue_61.R'
1 2022-07-05 10:27:11
```

**Comment by richelbilderbeek** [comment]
```
[richel@rackham2 ~]$ cat issue_61.log
# ...
1 2022-07-05 10:27:11
2 2022-07-05 10:32:09
3 2022-07-05 10:36:37
4 2022-07-05 10:41:04
[richel@rackham2 ~]$ ./nsphs_ml_qt/scripts_bianca/91_poll_jobs.sh 
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
          28166649      core issue_61   richel  R      16:38      1 r1224
```

**Comment by richelbilderbeek** [comment]
Typo in script (only measure at 1k epochs, now every 1k until 10k), restart:

```
[richel@rackham3 richel]$ sbatch nsphs_ml_qt/scripts_local/issue_61/start_issue_61.sh 
Submitted batch job 28170206
[richel@rackham3 richel]$ ./nsphs_ml_qt/scripts_bianca/91_poll_jobs.sh 
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
          28170206      core issue_61   richel  R       0:05      1 r1041
```

**Comment by richelbilderbeek** [comment]
I would say: GCAE does not converge:

[issue_61.csv](https://github.com/richelbilderbeek/nsphs_ml_qt/files/9072163/issue_61.csv)
![issue_61](https://user-images.githubusercontent.com/2098230/177995852-5bccd443-81c5-4706-b245-48a14ebc78e0.png)

What I did here:

 * Simuate a simple additive phenotype, with base -10, -9, -8, etc., so that phenotypic values are base +0, base + 0.5 and base +1.0
 * Run a regular GCAE

Instead that the error converges to zero, it does converges to something else.

**Comment by richelbilderbeek** [comment]
Video: [download (.ogv)](https://richelbilderbeek.nl/nsphs_nl_qt_issue_61.ogv) [YouTube](https://youtu.be/1O2mgY8A8-4)
